internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,139 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Tokener is an interface that defines the AccessToken method.
+type Tokener interface {
+	Token(context.Context) (string, error)
+	ForceRefresh(context.Context) (string, error)
+}
+
+type LazyToken struct {
+	// Url represents the URL used for acquiring the token.
+	Url string
+	// ClientId is the client ID used for authentication.
+	ClientId string
+	// ClientSecret is the client secret used for authentication.
+	ClientSecret string
+	// 	AccessToken string holds the currently cached token.
+	AccessToken string
+	// 	RefreshToken string token after refresh.
+	RefreshToken string
+	// Expiration stores the expiration time of the current token.
+	Expiration time.Time
+	// RefreshExpiration stores the expiration time of the refresh token.
+	RefreshExpiration time.Time
+	// TokenType is the type of the access token (e.g., "Bearer").
+	TokenType string
+	// NotBeforePolicy is the not-before policy of the access token.
+	NotBeforePolicy int
+	// Scope is the scope granted for the access token.
+	Scope string
```

Comment:
Just for the record, the three fields (token type, not before policy and scope) are not used. I am not against keeping this in memory, tho, there is no reason to keep it.